src/Shared/ChangeWaves.cs

```diff
@@ -88,121 +105,81 @@ internal static ChangeWaveConversionState ConversionState
+                return;
+            }
 
-            // If unset, enable all features.
-            if (DisabledWave.Length == 0 || DisabledWave.Equals(EnableAllFeatures, StringComparison.OrdinalIgnoreCase))
+            // Most common case, `MSBuildDisableFeaturesFromVersion` unset
+            if (string.IsNullOrEmpty(Traits.Instance.MSBuildDisableFeaturesFromVersion))
             {
                 ConversionState = ChangeWaveConversionState.Valid;
-                DisabledWave = ChangeWaves.EnableAllFeatures;
+                _cachedWave = ChangeWaves.EnableAllFeatures;
                 return;
             }
-
-            // If the version is of invalid format, log a warning and enable all features.
-            if (!Version.TryParse(DisabledWave, out changeWave))
+            else if (_cachedWave == null && !Version.TryParse(Traits.Instance.MSBuildDisableFeaturesFromVersion, out _cachedWave))
             {
                 ConversionState = ChangeWaveConversionState.InvalidFormat;
-                DisabledWave = ChangeWaves.EnableAllFeatures;
+                _cachedWave = ChangeWaves.EnableAllFeatures;
                 return;
             }
-            // If the version is 999.999, we're done.
-            else if (changeWave == EnableAllFeaturesAsVersion)
+            else if (_cachedWave == EnableAllFeatures || AllWaves.Contains(_cachedWave))
             {
                 ConversionState = ChangeWaveConversionState.Valid;
-                DisabledWave = changeWave.ToString();
                 return;
             }
-            // If the version is out of rotation, log a warning and clamp the value.
-            else if (changeWave < LowestWaveAsVersion)
+            else if (_cachedWave < LowestWave)
             {
                 ConversionState = ChangeWaveConversionState.OutOfRotation;
-                DisabledWave = LowestWave;
+                _cachedWave = LowestWave;
                 return;
             }
-            else if (changeWave > HighestWaveAsVersion)
+            else if (_cachedWave > HighestWave)
             {
                 ConversionState = ChangeWaveConversionState.OutOfRotation;
-                DisabledWave = HighestWave;
+                _cachedWave = HighestWave;
                 return;
             }
 
-            // Ensure it's set to an existing version within the current rotation
-            if (!AllWavesAsVersion.Contains(changeWave))
-            {
-                foreach (Version wave in AllWavesAsVersion)
-                {
-                    if (wave > changeWave)
-                    {
-                        ConversionState = ChangeWaveConversionState.Valid;
-                        DisabledWave = wave.ToString();
-                        return;
-                    }
-                }
-            }
-
+            // _cachedWave is somewhere between valid waves, find the next valid version.
+            _cachedWave = AllWaves.Where((x) => x > _cachedWave).First();
```

Comment:
```suggestion
            _cachedWave = AllWaves.First((x) => x > _cachedWave);
```
Still an unnecessary allocation, but this is clean, at least.

Reply:
This makes me wonder, would we ever advise setting the env var to something like `16.9`? I remember we mentioned something like this recently.